### PR TITLE
REF: delta_to_nanoseconds handle non-nano

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/_libs/tslibs/timestamps.pyx [hunk 1]
```diff
@@ -353,7 +353,17 @@ cdef class _Timestamp(ABCTimestamp):
             raise NotImplementedError(self._reso)
 
         if is_any_td_scalar(other):
-            nanos = delta_to_nanoseconds(other)
+            if (
+                is_timedelta64_object(other)
+                and get_datetime64_unit(other) == NPY_DATETIMEUNIT.NPY_FR_GENERIC
+            ):
+                # TODO: deprecate allowing this?  We only get here
```
Comment (mroeschke) [inline pandas/_libs/tslibs/timestamps.pyx:360]: Deprecate specifically adding a `np.timedelta64` object without a unit (and not the interval adding behavior)?

If so, +1.

Comment (jbrockmendel) [reply pandas/_libs/tslibs/timestamps.pyx:360]: correct

File: pandas/_libs/tslibs/timedeltas.pyx [hunk 2]
```diff
@@ -201,28 +201,76 @@ def ints_to_pytimedelta(ndarray m8values, box=False):
...
+            return n
+        else:
+            # _from_value_and_reso does not support Year, Month, or unit-less,
+            #  so we have special handling if speciifed
+            try:
+                td = Timedelta._from_value_and_reso(n, reso=in_reso)
+            except NotImplementedError:
+                if allow_year_month:
+                    td64 = ensure_td64ns(delta)
+                    return delta_to_nanoseconds(td64, reso=reso)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/_libs/tslibs/timedeltas.pyx:244]: sure, will update

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/_libs/tslibs/timedeltas.pyx:244]: wait, no. this is correct as-is

[the remark this answers is not in the document]
Comment (mroeschke) [reply pandas/_libs/tslibs/timedeltas.pyx:244]: Ah so this branch converts `delta` to ns no matter what to ensure the recursive call will exit appropriately the next time. Got it

**Review by mroeschke — APPROVED** [review]

**Pull request merged**